File: Week 27/cpp/closest_pair.cpp
/*
 * WEEK 27 - C++ ADVANCED TOPICS
 * Topic: Closest Pair of Points
 * File: closest_pair.cpp
 *
 * CONCEPT:
 *   Divide-and-conquer for the smallest pair distance among n points. Sort
 *   by x; recurse on each half; combine across the dividing line by
 *   examining only points within distance d of it, sorted by y.
 *
 * KEY POINTS:
 *   - Strip combine: each point compares with at most ~7 successors.
 *   - Base case: <= 3 points -> brute force.
 *   - O(n log n) total.
 *
 * ALGORITHM / APPROACH:
 *   sort by x
 *   solve(l, r):
 *     if small -> brute force
 *     d = min(solve(left), solve(right))
 *     build strip; sort by y; scan adjacent pairs while dy < d
 *     return d
 *
 * C++-SPECIFIC NOTES:
 *   - struct Point { double x, y; }
 *   - Use std::hypot for distance.
 *   - sort with lambda by x.
 *
 * DRY RUN / EXAMPLE:
 *   pts = {(2,3),(12,30),(40,50),(5,1),(12,10),(3,4)} -> ~1.4142 (between
 *   (2,3) and (3,4)).
 *
 * COMPLEXITY:
 *   Time O(n log n)   Space O(n)
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <climits>

using namespace std;

struct Point { double x, y; };

double dist(const Point& a, const Point& b) {
    return hypot(a.x - b.x, a.y - b.y);
}

double solve(vector<Point>& pts, int l, int r) {
    if (r - l < 3) {
        double best = 1e18;
        for (int i = l; i <= r; ++i)
            for (int j = i + 1; j <= r; ++j)
                best = min(best, dist(pts[i], pts[j]));
        return best;
    }
    int mid = (l + r) / 2;
    double midX = pts[mid].x;
    double d = min(solve(pts, l, mid), solve(pts, mid + 1, r));
    vector<Point> strip;
    for (int i = l; i <= r; ++i)
        if (fabs(pts[i].x - midX) < d) strip.push_back(pts[i]);
    sort(strip.begin(), strip.end(), [](auto& a, auto& b) { return a.y < b.y; });
    for (int i = 0; i < (int)strip.size(); ++i)
        for (int j = i + 1; j < (int)strip.size() && strip[j].y - strip[i].y < d; ++j)
            d = min(d, dist(strip[i], strip[j]));
    return d;
}

double closestPair(vector<Point> pts) {
    if (pts.size() < 2) return 1e18;
    sort(pts.begin(), pts.end(), [](auto& a, auto& b) { return a.x < b.x; });
    return solve(pts, 0, (int)pts.size() - 1);
}

double closestPairBrute(const vector<Point>& pts) {
    double best = 1e18;
    for (int i = 0; i < (int)pts.size(); ++i)
        for (int j = i + 1; j < (int)pts.size(); ++j)
            best = min(best, dist(pts[i], pts[j]));
    return best;
}

int main() {
    vector<Point> pts = {{2,3},{12,30},{40,50},{5,1},{12,10},{3,4}};
    printf("Closest pair (D&C):       %.4f\n", closestPair(pts));
    printf("Closest pair (brute force): %.4f\n", closestPairBrute(pts));
    return 0;
}

/*
 * NOTES
 * -----
 * Differences from Java:
 *   - C++ uses struct Point with doubles instead of int[].
 *   - std::hypot provides numerically stable Euclidean distance.
 *   - Brute-force baseline included for self-check.
 */

File: Week 27/rust/closest_pair.rs
// WEEK 27 - RUST ADVANCED TOPICS
// Topic: Closest Pair of Points
// File: closest_pair.rs
//
// CONCEPT:
//   Smallest pairwise Euclidean distance among n points. Sort by x, recurse,
//   then combine using a strip of width 2d sorted by y; each strip point
//   compares against at most ~7 successors -> O(n log n).
//
// KEY POINTS:
//   - Base case <= 3 -> brute force.
//   - Combine step is O(strip size).
//   - Use f64::hypot for stable distance.
//
// ALGORITHM / APPROACH:
//   sort by x
//   solve(l, r):
//     if small -> brute force
//     d = min(left, right)
//     build strip; sort by y; scan adjacent pairs while dy < d
//
// RUST-SPECIFIC NOTES:
//   - Pure recursion; the borrow checker is happy because we pass the
//     immutable slice through.
//   - sort_by with closure returning Ordering for f64 (use partial_cmp +
//     unwrap_or for total ordering).
//
// DRY RUN / EXAMPLE:
//   pts = [(2,3),(12,30),(40,50),(5,1),(12,10),(3,4)] -> ~1.4142.
//
// COMPLEXITY:
//   Time O(n log n)   Space O(n)

#[derive(Copy, Clone, Debug)]
pub struct Point { pub x: f64, pub y: f64 }

fn dist(a: Point, b: Point) -> f64 {
    (a.x - b.x).hypot(a.y - b.y)
}

fn solve(pts: &[Point], l: usize, r: usize) -> f64 {
    if r - l < 3 {
        let mut best = f64::INFINITY;
        for i in l..=r {
            for j in (i + 1)..=r {
                best = best.min(dist(pts[i], pts[j]));
            }
        }
        return best;
    }
    let mid = (l + r) / 2;
    let mid_x = pts[mid].x;
    let d_left = solve(pts, l, mid);
    let d_right = solve(pts, mid + 1, r);
    let mut d = d_left.min(d_right);
    let mut strip: Vec<Point> = pts[l..=r].iter().copied()
        .filter(|p| (p.x - mid_x).abs() < d)
        .collect();
    strip.sort_by(|a, b| a.y.partial_cmp(&b.y).unwrap());
    for i in 0..strip.len() {
        let mut j = i + 1;
        while j < strip.len() && strip[j].y - strip[i].y < d {
            d = d.min(dist(strip[i], strip[j]));
            j += 1;
        }
    }
    d
}

pub fn closest_pair(mut pts: Vec<Point>) -> f64 {
    if pts.len() < 2 { return f64::INFINITY; }
    pts.sort_by(|a, b| a.x.partial_cmp(&b.x).unwrap());
    solve(&pts, 0, pts.len() - 1)
}

pub fn closest_pair_brute(pts: &[Point]) -> f64 {
    let mut best = f64::INFINITY;
    for i in 0..pts.len() {
        for j in (i + 1)..pts.len() {
            best = best.min(dist(pts[i], pts[j]));
        }
    }
    best
}

fn main() {
    let pts = vec![
        Point{x:2.0,y:3.0},  Point{x:12.0,y:30.0}, Point{x:40.0,y:50.0},
        Point{x:5.0,y:1.0},  Point{x:12.0,y:10.0}, Point{x:3.0,y:4.0},
    ];
    println!("Closest pair (D&C):       {:.4}", closest_pair(pts.clone()));
    println!("Closest pair (brute force): {:.4}", closest_pair_brute(&pts));
}

// NOTES
// -----
// Differences from Java:
//   * f64::hypot replaces Math.sqrt + Math.pow for stable distance.
//   * partial_cmp().unwrap() yields a total Ord on f64 — assumes no NaN.
//   * Brute-force baseline included for self-check.
